### PR TITLE
feat: mitigate inverse relationship problems in swaps; more tests; clean up spot price

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

This PR is focused on truncations in the right direction as we square "square root price" in swaps.

While working on inverse relationship tests, I realized that some of the fractional residues due to intermediary math calculations are inevitable. Ref: https://github.com/osmosis-labs/osmosis/pull/4998#discussion_r1176463519

In swaps, what we want to guarantee is that we always round in the direction of the pool. As a result, here we introduce a method defined on the swap strategy that rounds in the desired direction based on swap strategy.

There are 2 drive-by changes here:
- Refactoring `CalculateSpotPrice` keeper method to use pool's spot price implementation instead of re-implementing on its own
- Adding more inverse relationship tick to price test vectors

## Brief Changelog
- add `math.SquareRoundUp` and `math.SquareRoundDown`
- add `SquareSqrtPrice` defined on each of the swap strategies and utilizing the methods above
- add tick to price inverse relationship tests
- refactor  `CalculateSpotPrice`

## Testing and Verifying

This change is already covered by existing tests

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/concentrated-liquidity/math/math.go [hunk 1]
```diff
@@ -173,3 +173,13 @@ func AddLiquidity(liquidityA, liquidityB sdk.Dec) (finalLiquidity sdk.Dec) {
 func MulRoundUp(a, b sdk.Dec) sdk.Dec {
 	return a.MulTruncate(b).Add(smallestDec)
 }
+
+// SquareRoundUp squares and rounds up at precision end.
+func SquareRoundUp(sqrtPrice sdk.Dec) sdk.Dec {
+	return sqrtPrice.MulRoundUp(sqrtPrice)
+}
+
+// quareTruncate squares and truncates at precision end.
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/math/math.go:182]:
```suggestion
// SquareTruncate squares and truncates at precision end.
```

File: x/concentrated-liquidity/math/tick_test.go [hunk 2]
```diff
@@ -309,7 +310,7 @@ func (suite *ConcentratedMathTestSuite) TestTickToSqrtPricePriceToTick_InverseRe
 			price:        types.MaxSpotPrice,
 			tickExpected: sdk.NewInt(types.MaxTick).String(),
 		},
-		"max spot price - smallest dec": {
+		"max spot price - smallest price delta given exponent at price one of 6": {
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/math/tick_test.go:313]:
```suggestion
		"max spot price - smallest price delta given exponent at price one of -6": {
```

File: x/concentrated-liquidity/swapstrategy/one_for_zero.go [hunk 3]
```diff
@@ -221,3 +221,13 @@ func (s oneForZeroStrategy) ValidateSqrtPrice(sqrtPrice, currentSqrtPrice sdk.De
 	}
 	return nil
 }
+
+// SquarePrice returns the square of the sqrt price (price).
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/swapstrategy/one_for_zero.go:225]:
```suggestion
// SquareSqrtPrice returns the square of the sqrt price (price).
```

File: x/concentrated-liquidity/swapstrategy/swap_strategy.go [hunk 4]
```diff
@@ -72,6 +72,10 @@ type swapStrategy interface {
 	// and the min/max sqrt price on the other side.
 	// See oneForZeroStrategy or zeroForOneStrategy for implementation details.
 	ValidateSqrtPrice(sqrtPriceLimit, currentSqrtPrice sdk.Dec) error
+	// SquarePrice returns the square of the sqrt price (price).
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/swapstrategy/swap_strategy.go:75]:
```suggestion
	// SquareSqrtPrice returns the square of the sqrt price (price).
```

**Review by czarcas7ic — APPROVED** [review]

**Review by AlpinYukseloglu — APPROVED** [review]
Nice work, I really like the abstractions you set up here

**Pull request merged**